// @visactor/vmind/esm/atom/chartGenerator/spec/transformers/circlePacking.js
import { color, data } from "./common";

export const bubbleCirclePackingData = context => {
    const {dataTable: dataTable, spec: spec, cell: cell} = context;
    return cell.size && dataTable.forEach((data => {
        data.value = data[cell.size], "value" !== cell.size && delete data[cell.size];
    })), {
        spec: spec
    };
};

export const bubbleCirclePackingField = context => {
    const {cell: cell, spec: spec} = context;
    return spec.categoryField = cell.color || cell.x, cell.size && (spec.valueField = cell.size), 
    {
        spec: spec
    };
};

export const bubbleCirclePackingDisplayConf = context => {
    const {spec: spec} = context;
    return spec.drill = !0, spec.layoutPadding = 5, spec.animationEnter = {
        easing: "cubicInOut"
    }, spec.animationExit = {
        easing: "cubicInOut"
    }, spec.animationUpdate = {
        easing: "cubicInOut"
    }, {
        spec: spec
    };
};

export const pipelineBubbleCirclePacking = [ bubbleCirclePackingData, data, color, bubbleCirclePackingField, bubbleCirclePackingDisplayConf ];
