// @visactor/vmind/esm/atom/chartGenerator/spec/transformers/wordcloud.js
import { isValidDataTable } from "../../../../utils/dataTable";

import { WORDCLOUD_NUM_LIMIT } from "../constants";

import { color } from "./common";

export const wordCloudData = context => {
    const {dataTable: dataTable, spec: spec} = context;
    return spec.data = {
        id: "data",
        values: isValidDataTable(dataTable) ? dataTable.slice(0, WORDCLOUD_NUM_LIMIT) : []
    }, {
        spec: spec
    };
};

export const wordCloudField = context => {
    const {cell: cell, spec: spec} = context;
    return spec.nameField = cell.color, cell.size && (spec.valueField = cell.size), 
    spec.seriesField = spec.nameField, {
        spec: spec
    };
};

export const wordCloudDisplayConf = context => {
    const {spec: spec, chartTheme: chartTheme} = context;
    return chartTheme || (spec.fontSizeRange = [ 20, 50 ], spec.fontWeightRange = [ 800, 800 ]), 
    {
        spec: spec
    };
};

export const pipelineWordCloud = [ wordCloudData, color, wordCloudField, wordCloudDisplayConf ];
