// @visactor/vmind/cjs/atom/chartGenerator/spec/transformers/circlePacking.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.pipelineBubbleCirclePacking = exports.bubbleCirclePackingDisplayConf = exports.bubbleCirclePackingField = exports.bubbleCirclePackingData = void 0;

const common_1 = require("./common"), bubbleCirclePackingData = context => {
    const {dataTable: dataTable, spec: spec, cell: cell} = context;
    return cell.size && dataTable.forEach((data => {
        data.value = data[cell.size], "value" !== cell.size && delete data[cell.size];
    })), {
        spec: spec
    };
};

exports.bubbleCirclePackingData = bubbleCirclePackingData;

const bubbleCirclePackingField = context => {
    const {cell: cell, spec: spec} = context;
    return spec.categoryField = cell.color || cell.x, cell.size && (spec.valueField = cell.size), 
    {
        spec: spec
    };
};

exports.bubbleCirclePackingField = bubbleCirclePackingField;

const bubbleCirclePackingDisplayConf = context => {
    const {spec: spec} = context;
    return spec.drill = !0, spec.layoutPadding = 5, spec.animationEnter = {
        easing: "cubicInOut"
    }, spec.animationExit = {
        easing: "cubicInOut"
    }, spec.animationUpdate = {
        easing: "cubicInOut"
    }, {
        spec: spec
    };
};

exports.bubbleCirclePackingDisplayConf = bubbleCirclePackingDisplayConf, exports.pipelineBubbleCirclePacking = [ exports.bubbleCirclePackingData, common_1.data, common_1.color, exports.bubbleCirclePackingField, exports.bubbleCirclePackingDisplayConf ];
