// @visactor/vmind/cjs/atom/chartGenerator/spec/transformers/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.sequenceData = exports.wordCloudData = exports.sankeyData = exports.funnelData = exports.legend = exports.data = void 0;

var common_1 = require("./common");

Object.defineProperty(exports, "data", {
    enumerable: !0,
    get: function() {
        return common_1.data;
    }
}), Object.defineProperty(exports, "legend", {
    enumerable: !0,
    get: function() {
        return common_1.discreteLegend;
    }
});

var funnel_1 = require("./funnel");

Object.defineProperty(exports, "funnelData", {
    enumerable: !0,
    get: function() {
        return funnel_1.funnelData;
    }
});

var sankey_1 = require("./sankey");

Object.defineProperty(exports, "sankeyData", {
    enumerable: !0,
    get: function() {
        return sankey_1.sankeyData;
    }
});

var wordcloud_1 = require("./wordcloud");

Object.defineProperty(exports, "wordCloudData", {
    enumerable: !0,
    get: function() {
        return wordcloud_1.wordCloudData;
    }
});

var rankingBar_1 = require("./rankingBar");

Object.defineProperty(exports, "sequenceData", {
    enumerable: !0,
    get: function() {
        return rankingBar_1.sequenceData;
    }
});
