// @visactor/vmind/cjs/atom/chartGenerator/spec/transformers/line.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.pipelineLine = exports.displayConfLine = exports.cartesianLine = exports.colorLine = void 0;

const field_1 = require("../../../../utils/field"), constants_1 = require("../constants"), common_1 = require("./common"), base_1 = require("../../../../types/base"), cartesian_1 = require("./cartesian"), colorLine = context => {
    const {colors: colors, spec: spec, chartTheme: chartTheme} = context;
    return chartTheme || (colors && colors.length > 0 ? spec.color = colors : (spec.color = constants_1.LINEAR_COLOR_THEMES.map((c => ({
        gradient: "linear",
        x0: 0,
        y0: .5,
        x1: 1,
        y1: .5,
        stops: [ {
            offset: 0,
            color: c[0]
        }, {
            offset: 1,
            color: c[1]
        } ]
    }))), spec.point = {
        style: {}
    })), {
        spec: spec
    };
};

exports.colorLine = colorLine;

const cartesianLine = context => {
    const {cell: cell, spec: spec, fieldInfo: fieldInfo} = context, cellNew = Object.assign({}, cell);
    if (spec.xField = cell.x, spec.yField = cell.y, cell.color) spec.seriesField = cell.color; else {
        const remainedFields = fieldInfo.filter((({fieldName: fieldName}) => !spec.xField.includes(fieldName) && spec.yField !== fieldName)), colorField = (0, 
        field_1.getFieldByDataType)(remainedFields, [ base_1.DataType.STRING, base_1.DataType.DATE ]);
        colorField && (spec.seriesField = colorField.fieldName, cellNew.color = colorField.fieldName);
    }
    return {
        spec: spec,
        cell: cellNew
    };
};

exports.cartesianLine = cartesianLine;

const displayConfLine = context => {
    const {spec: spec, chartTheme: chartTheme} = context;
    return chartTheme || (spec.line = {
        style: {}
    }), {
        spec: spec
    };
};

exports.displayConfLine = displayConfLine, exports.pipelineLine = [ common_1.data, exports.colorLine, exports.cartesianLine, cartesian_1.seriesField, cartesian_1.axis, common_1.discreteLegend, exports.displayConfLine ];
