// @visactor/vmind/cjs/atom/chartGenerator/spec/transformers/linearProgress.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.pipelineLinearProgress = exports.linearProgressStyle = exports.linearProgressAxes = exports.linearProgressField = void 0;

const common_1 = require("./common"), linearProgressField = context => {
    const {cell: cell, spec: spec} = context;
    return spec.xField = cell.y, spec.yField = cell.x, cell.color && (spec.seriesField = cell.color), 
    spec.cornerRadius = 20, {
        spec: spec
    };
};

exports.linearProgressField = linearProgressField;

const linearProgressAxes = context => {
    const {cell: cell, spec: spec} = context, hasSingleData = spec.data.values && 1 === spec.data.values.length;
    return spec.axes = [ {
        orient: "left",
        type: "band",
        domainLine: {
            visible: !1
        },
        tick: {
            visible: !1
        },
        label: {
            formatMethod: hasSingleData ? val => `${cell.x}: ${val}` : null,
            style: {
                fontSize: 16
            }
        }
    }, {
        orient: "bottom",
        type: "linear",
        visible: !0,
        grid: {
            visible: !1
        },
        label: {
            formatMethod: val => val >= 0 && val <= 1 ? 100 * val + "%" : val,
            flush: !0
        }
    } ], {
        spec: spec
    };
};

exports.linearProgressAxes = linearProgressAxes;

const linearProgressStyle = context => {
    const {spec: spec} = context;
    return spec.progress = Object.assign(Object.assign({}, spec.progress), {
        style: {}
    }), {
        spec: spec
    };
};

exports.linearProgressStyle = linearProgressStyle, exports.pipelineLinearProgress = [ common_1.data, common_1.color, exports.linearProgressField, exports.linearProgressAxes, exports.linearProgressStyle ];
