// @visactor/vmind/cjs/atom/chartGenerator/spec/transformers/map.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.pipelineMapChart = exports.mapDisplayConf = exports.mapField = exports.basemap = void 0;

const common_1 = require("./common"), basemap = context => {
    const {basemapOption: basemapOption, spec: spec} = context;
    return basemapOption.regionProjectType ? spec.region = [ {
        roam: !0,
        projection: {
            type: basemapOption.regionProjectType
        },
        coordinate: basemapOption.regionCoordinate
    } ] : spec.region = [ {
        roam: !0,
        coordinate: basemapOption.regionCoordinate
    } ], spec.map = "map", {
        spec: spec
    };
};

exports.basemap = basemap;

const mapField = context => {
    const {spec: spec, cell: cell} = context;
    return spec.nameField = cell.color, spec.valueField = cell.size, spec.nameProperty = cell.color, 
    {
        spec: spec
    };
};

exports.mapField = mapField;

const mapDisplayConf = context => {
    const {spec: spec, cell: cell} = context;
    return spec.legends = [ {
        visible: !0,
        type: "color",
        field: cell.size,
        orient: "bottom",
        position: "start"
    } ], spec.area = {
        style: {
            fill: {
                field: cell.size,
                scale: "color",
                changeDomain: "replace"
            }
        }
    }, {
        spec: spec
    };
};

exports.mapDisplayConf = mapDisplayConf, exports.pipelineMapChart = [ exports.basemap, common_1.color, common_1.arrayData, exports.mapField, exports.mapDisplayConf ];
