// @visactor/vmind/cjs/atom/chartGenerator/spec/transformers/rose.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.pipelineRose = exports.roseAxis = exports.roseField = void 0;

const common_1 = require("./common"), roseField = context => {
    var _a;
    const {cell: cell, spec: spec} = context;
    return spec.valueField = cell.radius || cell.angle, cell.color && (spec.categoryField = null !== (_a = cell.category) && void 0 !== _a ? _a : cell.color, 
    spec.seriesField = cell.color), spec.outerRadius = .8, spec.innerRadius = .2, {
        spec: spec
    };
};

exports.roseField = roseField;

const roseAxis = context => {
    const {spec: spec} = context;
    return spec.axes = [ {
        orient: "angle",
        domainLine: {
            visible: !1
        },
        grid: {
            visible: !1,
            alignWithLabel: !1
        },
        label: {
            visible: !0
        }
    }, {
        orient: "radius",
        grid: {
            visible: !1,
            smooth: !0
        }
    } ], {
        spec: spec
    };
};

exports.roseAxis = roseAxis, exports.pipelineRose = [ common_1.data, common_1.color, exports.roseField, exports.roseAxis, common_1.discreteLegend, common_1.commonLabel ];
