// @visactor/vmind/cjs/atom/chartGenerator/spec/transformers/venn.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.pipelineVenn = exports.registerChart = exports.vennField = exports.vennData = void 0;

const vchart_1 = require("@visactor/vchart"), common_1 = require("./common"), vennData = context => {
    const {dataTable: dataTable, spec: spec, cell: cell} = context, id2dataMap = {}, setsField = cell.color[0], nameField = cell.color[1];
    return dataTable.forEach((data => {
        id2dataMap[data[setsField]] ? id2dataMap[data[setsField]].sets.push(data[nameField]) : id2dataMap[data[setsField]] = {
            sets: [ data[nameField] ],
            value: data[cell.size]
        };
    })), spec.data = {
        values: Object.values(id2dataMap)
    }, {
        spec: spec
    };
};

exports.vennData = vennData;

const vennField = context => {
    const {spec: spec} = context;
    return spec.valueField = "value", spec.categoryField = "sets", spec.seriesField = "sets", 
    {
        spec: spec
    };
};

exports.vennField = vennField;

const registerChart = context => {
    const {spec: spec} = context;
    return "venn" === spec.type && (0, vchart_1.registerVennChart)(), {
        spec: spec
    };
};

exports.registerChart = registerChart, exports.pipelineVenn = [ exports.registerChart, exports.vennData, common_1.color, exports.vennField, common_1.discreteLegend ];
