// @visactor/vmind/cjs/atom/chartGenerator/spec/transformers/waterfall.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.pipelineWaterfall = exports.waterfallStackLabel = exports.waterfallAxes = exports.waterfallField = void 0;

const common_1 = require("./common"), waterfallField = context => {
    const {cell: cell, spec: spec} = context;
    return spec.xField = cell.x, spec.yField = cell.y, spec.total = {
        type: "end",
        text: "总计"
    }, {
        spec: spec
    };
};

exports.waterfallField = waterfallField;

const waterfallAxes = context => {
    const {spec: spec} = context;
    return spec.axes = [ {
        orient: "left",
        title: {
            visible: !0,
            text: "favorability"
        },
        label: {
            formatMethod: v => v + "%"
        }
    }, {
        orient: "bottom",
        label: {
            visible: !0
        },
        type: "band",
        paddingInner: .4,
        title: {
            visible: !0,
            text: "date"
        }
    } ], {
        spec: spec
    };
};

exports.waterfallAxes = waterfallAxes;

const waterfallStackLabel = context => {
    const {spec: spec} = context;
    return spec.stackLabel = {
        valueType: "absolute",
        formatMethod: text => text + "%"
    }, {
        spec: spec
    };
};

exports.waterfallStackLabel = waterfallStackLabel, exports.pipelineWaterfall = [ common_1.data, common_1.color, exports.waterfallField, exports.waterfallAxes, exports.waterfallStackLabel, common_1.discreteLegend ];
