// @visactor/vmind/cjs/atom/chartGenerator/spec/transformers/wordcloud.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.pipelineWordCloud = exports.wordCloudDisplayConf = exports.wordCloudField = exports.wordCloudData = void 0;

const dataTable_1 = require("../../../../utils/dataTable"), constants_1 = require("../constants"), common_1 = require("./common"), wordCloudData = context => {
    const {dataTable: dataTable, spec: spec} = context;
    return spec.data = {
        id: "data",
        values: (0, dataTable_1.isValidDataTable)(dataTable) ? dataTable.slice(0, constants_1.WORDCLOUD_NUM_LIMIT) : []
    }, {
        spec: spec
    };
};

exports.wordCloudData = wordCloudData;

const wordCloudField = context => {
    const {cell: cell, spec: spec} = context;
    return spec.nameField = cell.color, cell.size && (spec.valueField = cell.size), 
    spec.seriesField = spec.nameField, {
        spec: spec
    };
};

exports.wordCloudField = wordCloudField;

const wordCloudDisplayConf = context => {
    const {spec: spec, chartTheme: chartTheme} = context;
    return chartTheme || (spec.fontSizeRange = [ 20, 50 ], spec.fontWeightRange = [ 800, 800 ]), 
    {
        spec: spec
    };
};

exports.wordCloudDisplayConf = wordCloudDisplayConf, exports.pipelineWordCloud = [ exports.wordCloudData, common_1.color, exports.wordCloudField, exports.wordCloudDisplayConf ];
